src/hooks/use-should-position-above.ts

```diff
@@ -0,0 +1,49 @@
+import { useLayoutEffect, useState } from 'preact/hooks';
+
+/**
+ * Determines if a popping element (typically a dialog/popover of some sort)
+ * should be positioned above another reference component, based on the sizes of
+ * those two components, and the available viewport space above and below the
+ * reference component.
+ *
+ * By default, we prefer the popping component to be positioned below the
+ * reference component, and only if there's enough space above, and not enough
+ * space below, we will prefer it to be positioned above.
+ */
+export function useShouldPositionAbove(
+  referenceComponent: HTMLElement | null,
+  poppingComponent: HTMLElement | null,
```

Comment:
I don't love the `referenceComponent` and `poppingComponent` names, so I'm open to suggestions here.